python/iceberg/core/filesystem/s3_filesystem.py

```diff
@@ -0,0 +1,274 @@
+from botocore.session import get_session
+from retrying import retry
+
+from .file_status import FileStatus
+from .file_system import FileSystem
+
+_logger = logging.getLogger(__name__)
+
+
+S3_CLIENT = dict()
+BOTO_STS_CLIENT = boto3.client('sts')
+ROLE_ARN = None
+AUTOREFRESH_SESSION = None
+
+
+@retry(wait_incrementing_start=100, wait_exponential_multiplier=4,
+       wait_exponential_max=5000, stop_max_delay=600000, stop_max_attempt_number=7)
+def get_s3(obj="resource"):
+    global AUTOREFRESH_SESSION
+    global ROLE_ARN
+    if ROLE_ARN not in S3_CLIENT:
+        S3_CLIENT[ROLE_ARN] = dict()
+
+    if ROLE_ARN == "default":
+        if AUTOREFRESH_SESSION is None:
+            AUTOREFRESH_SESSION = boto3.Session()
+        S3_CLIENT["default"]["resource"] = AUTOREFRESH_SESSION.resource('s3')
+        S3_CLIENT["default"]["client"] = AUTOREFRESH_SESSION.client('s3')
+    else:
+        if AUTOREFRESH_SESSION is None:
+            sess = get_session()
+            sess._credentials = RefreshableCredentials.create_from_metadata(metadata=refresh_sts_session_keys(),
+                                                                            refresh_using=refresh_sts_session_keys,
+                                                                            method="sts-assume-role")
+            AUTOREFRESH_SESSION = boto3.Session(botocore_session=sess)
+
+        S3_CLIENT[ROLE_ARN]["resource"] = AUTOREFRESH_SESSION.resource("s3")
+        S3_CLIENT[ROLE_ARN]["client"] = AUTOREFRESH_SESSION.client("s3")
+
+    return S3_CLIENT.get(ROLE_ARN, dict()).get(obj)
+
+
+def refresh_sts_session_keys():
+    params = {"RoleArn": ROLE_ARN,
+              "RoleSessionName": "iceberg_python_client_{}".format(int(time.time() * 1000.00))}
+
+    sts_creds = BOTO_STS_CLIENT.assume_role(**params).get("Credentials")
+    credentials = {"access_key": sts_creds.get("AccessKeyId"),
+                   "secret_key": sts_creds.get("SecretAccessKey"),
+                   "token": sts_creds.get("SessionToken"),
+                   "expiry_time": sts_creds.get("Expiration").isoformat()}
+    return credentials
+
+
+def url_to_bucket_key_name_tuple(url):
+    parsed_url = urlparse(url)
+    return parsed_url.netloc, parsed_url.path[1:], parsed_url.path.split("/")[-1]
+
+
+class S3FileSystem(FileSystem):
+    fs_inst = None
+
+    @staticmethod
+    def get_instance():
+        if S3FileSystem.fs_inst is None:
+            S3FileSystem()
+        return S3FileSystem.fs_inst
+
+    def __init__(self):
+        if S3FileSystem.fs_inst is None:
+            S3FileSystem.fs_inst = self
+
+    def set_conf(self, conf):
+        global ROLE_ARN
+        if conf is not None:
+            ROLE_ARN = conf.get("hive.aws_iam_role", "default")
+
+    def exists(self, path):
+        bucket, key, _ = url_to_bucket_key_name_tuple(S3FileSystem.normalize_s3_path(path))
+
+        res = get_s3(obj="client").list_objects_v2(Bucket=bucket,
```

Comment:
We should probably be doing a get object metadata request (e.g. HEAD request though I don't know what the right boto3 call would map to that) for existence check.  I'm not even certain this is correct as the list call will also hit for any prefix that matches the path.

Reply:
Looks like I'm already doing the head request in the info method, so I'll just call that and catch if it 404's and make that false and true if it succeeds